Problem with datetime properties with null values

## Expected Behavior (Mandatory)

Create a dataframe based on the cypher query

## Actual Behavior (Mandatory)

Works fine on Databricks Runtime 9.1 LTS

Breaks on Databricks Runtime 12.2 LTS - Error Message->IOException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.ZonedDateTime` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling
Caused by: InvalidDefinitionException: Java 8 date/time type `java.time.ZonedDateTime` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling

## How to Reproduce the Problem

Use any cypher query that returns more than one datetime, like: datetime(o.submittedDate) as submittedDateABC, o.isOnboardingSubmitted as isOnboardingSubmitted, datetime(o.triggeredDate) as triggeredDateABC, datetime(o.sentBackDate) as sentBackDateABC, o.finalActionTakenUsername as finalActionTakenUsername, datetime(o.onboardingFinalActionTakenDate) as onboardingFinalActionTakenDate, o.statusid as statusId

## Specifications (Mandatory)

Currently used versions

### Versions

- Spark: 3.3.2
- Scala: 2.12.15
- Neo4j: 4.4.15
- Neo4j Connector: 5.0.1

## Additional information

 Spark Cloud provider: Databricks 12.2LTS (https://docs.databricks.com/release-notes/runtime/12.2.html)

--- a/common/src/main/scala/org/neo4j/spark/reader/BasePartitionReader.scala
+++ b/common/src/main/scala/org/neo4j/spark/reader/BasePartitionReader.scala
@@ -12,7 +12,6 @@ import org.neo4j.spark.util.{DriverCache, Neo4jOptions, Neo4jUtil, QueryType}
 import java.io.IOException
 import java.util
 import scala.collection.JavaConverters._
-import scala.collection.mutable
 
 abstract class BasePartitionReader(private val options: Neo4jOptions,
                                    private val filters: Array[Filter],

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
@@ -20,6 +20,7 @@ import org.slf4j.Logger
 
 import java.time._
 import java.time.format.DateTimeFormatter
+import java.time.temporal.Temporal
 import java.util.Properties
 import scala.collection.JavaConverters._
 
@@ -68,6 +69,11 @@ object Neo4jUtil {
     override def serialize(entity: Entity,
                            jsonGenerator: JsonGenerator,
                            serializerProvider: SerializerProvider): Unit = jsonGenerator.writeObject(entity.toMap)
+  })
+  module.addSerializer(classOf[Temporal], new JsonSerializer[Temporal]() {
+    override def serialize(entity: Temporal,
+                           jsonGenerator: JsonGenerator,
+                           serializerProvider: SerializerProvider): Unit = jsonGenerator.writeRaw(entity.toString)
   })
   mapper.registerModule(module)
 

--- a/spark-3/src/test/scala/org/neo4j/spark/DataSourceReaderTSE.scala
+++ b/spark-3/src/test/scala/org/neo4j/spark/DataSourceReaderTSE.scala
@@ -1702,6 +1702,74 @@ class DataSourceReaderTSE extends SparkConnectorScalaBaseTSE {
     assertEquals(Set("target.name", "target.id"), df.columns.toSet)
   }
 
+  @Test
+  def test531(): Unit = {
+    val dataFrame = ss.read.format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("query",
+        """
+          |UNWIND [
+          |  {first: '2022-06-14T10:02:28.192Z', second: null},
+          |  {first: '2022-06-15T10:02:28.192Z', second: '2022-06-16T10:02:28.192Z'}
+          |]AS event
+          |RETURN datetime(event.first) AS first, datetime(event.second) AS second
+          |""".stripMargin)
+      .load()
+
+    assertEquals(
+      StructType(Array(
+        StructField("first", DataTypes.TimestampType),
+        StructField("second", DataTypes.StringType)
+      )),
+      dataFrame.schema
+    )
+
+    assertEquals(
+      List(
+        (Timestamp.from(OffsetDateTime.parse("2022-06-14T10:02:28.192Z").toInstant), null),
+        (Timestamp.from(OffsetDateTime.parse("2022-06-15T10:02:28.192Z").toInstant), "2022-06-16T10:02:28.192Z")
+      ),
+      dataFrame.collect()
+        .map(r => (r.getTimestamp(0), r.getString(1)))
+        .toList
+    )
+  }
+
+  @Test
+  def test531WithSchema(): Unit = {
+    val schema = StructType(Array(
+      StructField("first", DataTypes.TimestampType),
+      StructField("second", DataTypes.TimestampType),
+    ))
+    // dataFrameWithSchema must with the proper data type
+    val dataFrameWithSchema = ss.read.format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("query",
+        """
+          |UNWIND [
+          |  {first: '2022-06-14T10:02:28.192Z', second: null},
+          |  {first: '2022-06-15T10:02:28.192Z', second: '2022-06-16T10:02:28.192Z'}
+          |] AS event
+          |RETURN datetime(event.first) AS first, datetime(event.second) AS second
+          |""".stripMargin)
+      .schema(schema)
+      .load()
+    assertEquals(
+      schema,
+      dataFrameWithSchema.schema
+    )
+    assertEquals(
+      List(
+        (Timestamp.from(OffsetDateTime.parse("2022-06-14T10:02:28.192Z").toInstant), null),
+        (Timestamp.from(OffsetDateTime.parse("2022-06-15T10:02:28.192Z").toInstant),
+          Timestamp.from(OffsetDateTime.parse("2022-06-16T10:02:28.192Z").toInstant))
+      ),
+      dataFrameWithSchema.collect()
+        .map(r => (r.getTimestamp(0), r.getTimestamp(1)))
+        .toList
+    )
+  }
+
   private def initTest(query: String): DataFrame = {
     SparkConnectorScalaSuiteIT.session()
       .writeTransaction(

